crates/tests/src/vm_host_env/tx.rs

```diff
@@ -521,3 +521,293 @@ mod native_tx_host_env {
+    }
+
+    proptest! {
+        #[test]
+        fn test_tx_iter_prefix_cannot_panic(
+            setup in arb_test_setup(),
+        ) {
+            test_tx_iter_prefix_cannot_panic_aux(setup)
+        }
+    }
+
+    fn test_tx_iter_prefix_cannot_panic_aux(setup: TestSetup) {
+        // dbg!(&setup);
+
+        let mut test_env = TestTxEnv::default();
+        let tx_env = setup_host_env(&setup, &mut test_env);
+
+        // Can fail, but must not panic
+        let _res = host_env::tx_iter_prefix(
+            &tx_env,
+            setup.key_memory_ptr,
+            setup.key_len(),
+        );
+        let _res = host_env::tx_iter_next(
+            &tx_env,
+            // This field is not used for anything else in this test
+            setup.val_memory_ptr,
+        );
+    }
+
+    fn setup_host_env(
+        setup: &TestSetup,
+        test_env: &mut TestTxEnv,
+    ) -> TxVmEnv<
+        'static,
+        wasm::memory::WasmMemory,
+        MockDB,
+        Sha256Hasher,
+        WasmCacheRwAccess,
+    > {
+        if setup.write_to_storage {
+            // Write the key-val to storage which may affect `tx_read` execution
+            // path
+            let _res =
+                test_env.wl_storage.storage.write(&setup.key, &setup.val);
+        }
+        if setup.write_to_wl {
+            // Write the key-val to write log which may affect `tx_read`
+            // execution path
+            let _res = test_env
+                .wl_storage
+                .write_log
+                .write(&setup.key, setup.val.clone());
+        }
+
+        let TestTxEnv {
+            wl_storage,
+            iterators,
+            verifiers,
+            gas_meter,
+            sentinel,
+            result_buffer,
+            tx_index,
+            vp_wasm_cache,
+            vp_cache_dir: _,
+            tx_wasm_cache,
+            tx_cache_dir: _,
+            tx,
+        } = test_env;
+
+        let tx_env = vm::host_env::testing::tx_env_with_wasm_memory(
+            &wl_storage.storage,
+            &mut wl_storage.write_log,
+            iterators,
+            verifiers,
+            gas_meter,
+            sentinel,
+            tx,
+            tx_index,
+            result_buffer,
+            vp_wasm_cache,
+            tx_wasm_cache,
+        );
+
+        if setup.write_to_memory {
+            let key_bytes = setup.key_bytes();
+            // Write the key-val to memory which may affect `tx_read` execution
+            // path. Can fail, but must not panic
+            let _res =
+                tx_env.memory.write_bytes(setup.key_memory_ptr, key_bytes);
+        }
+
+        tx_env
+    }
+
+    fn arb_test_setup() -> impl Strategy<Value = TestSetup> {
+        (
+            any::<bool>(),
+            any::<bool>(),
+            any::<bool>(),
+            namada::types::storage::testing::arb_key(),
+            arb_u64(),
+            arb_u64(),
+            any::<Vec<u8>>(),
+            arb_u64(),
```

Comment:
idk how proptest shrinks results, but half of `u64` values will be invalid. depends on how biased we want the tests to be (over failure or success). maybe it doesn't matter here.

Reply:
I've added some weights to the strategy below to change to be more likely to use the valid range. We still do want the invalid values here too though

Reply:
[6e6e0d6](https://github.com/anoma/namada/pull/2401/commits/6e6e0d6523887c0e7030c94c3eb4084029b6bd7d)